File: clients/android/StellarChat/app/src/main/java/com/stellarmls/chat/crypto/MediaCrypto.kt
package com.stellarmls.chat.crypto

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.media.MediaMetadataRetriever
import android.net.Uri
import java.io.ByteArrayOutputStream
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Per-file AES-256-GCM encryption for media attachments.
 * Each file gets a fresh random key to avoid reusing the group key
 * for large binary payloads, minimizing AES-GCM nonce-collision risk.
 */
object MediaCrypto {
    /** Maximum compressed image size in bytes (2 MB). */
    const val MAX_IMAGE_BYTES = 2_000_000
    /** Maximum video size in bytes before encryption (50 MB). */
    const val MAX_VIDEO_BYTES = 50_000_000
    /** Maximum thumbnail dimension in pixels. */
    const val THUMBNAIL_MAX_DIMENSION = 200
    /** Thumbnail JPEG compression quality (0-100). */
    const val THUMBNAIL_QUALITY = 40

    private const val NONCE_SIZE = 12
    private const val TAG_SIZE = 16

    // -- Encrypt / Decrypt --

    /**
     * Encrypt data with a fresh random AES-256-GCM key.
     * Returns Pair(encryptedBlob, key) where encryptedBlob = nonce(12) || ciphertext || tag(16).
     */
    fun encryptMedia(data: ByteArray): Pair<ByteArray, ByteArray> {
        val key = ByteArray(32)
        SecureRandom().nextBytes(key)
        val encrypted = encryptMedia(data, key)
        return Pair(encrypted, key)
    }

    /**
     * Encrypt data with a provided AES-256-GCM key.
     * Returns combined format: nonce(12) || ciphertext || tag(16).
     */
    fun encryptMedia(data: ByteArray, key: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = SecretKeySpec(key, "AES")
        cipher.init(Cipher.ENCRYPT_MODE, keySpec)
        val nonce = cipher.iv // 12 bytes generated automatically
        val ciphertextWithTag = cipher.doFinal(data)

        // Combined format: nonce || ciphertext || tag
        // Java GCM appends tag to ciphertext, so ciphertextWithTag = ct + tag already
        return nonce + ciphertextWithTag
    }

    /**
     * Decrypt combined-format blob: nonce(12) || ciphertext || tag(16).
     */
    fun decryptMedia(encryptedBlob: ByteArray, key: ByteArray): ByteArray {
        require(encryptedBlob.size > NONCE_SIZE + TAG_SIZE) { "Encrypted blob too short" }

        val nonce = encryptedBlob.copyOfRange(0, NONCE_SIZE)
        val ciphertextWithTag = encryptedBlob.copyOfRange(NONCE_SIZE, encryptedBlob.size)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val keySpec = SecretKeySpec(key, "AES")
        val gcmSpec = GCMParameterSpec(TAG_SIZE * 8, nonce)
        cipher.init(Cipher.DECRYPT_MODE, keySpec, gcmSpec)
        return cipher.doFinal(ciphertextWithTag)
    }

    // -- Image Processing --

    /**
     * Compress an image to JPEG within [maxBytes].
     * Progressively reduces quality until the size limit is met.
     */
    fun compressImage(imageData: ByteArray, maxBytes: Int = MAX_IMAGE_BYTES): ByteArray? {
        val bitmap = BitmapFactory.decodeByteArray(imageData, 0, imageData.size) ?: return null
        var quality = 70
        while (quality >= 10) {
            val output = ByteArrayOutputStream()
            bitmap.compress(Bitmap.CompressFormat.JPEG, quality, output)
            val result = output.toByteArray()
            if (result.size <= maxBytes) {
                bitmap.recycle()
                return result
            }
            quality -= 10
        }
        // Last resort: lowest quality
        val output = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 10, output)
        bitmap.recycle()
        return output.toByteArray()
    }

    /**
     * Generate a small JPEG thumbnail.
     */
    fun generateThumbnail(
        imageData: ByteArray,
        maxDimension: Int = THUMBNAIL_MAX_DIMENSION
    ): ByteArray? {
        val bitmap = BitmapFactory.decodeByteArray(imageData, 0, imageData.size) ?: return null
        val scale: Float = if (bitmap.width > bitmap.height) {
            maxDimension.toFloat() / bitmap.width
        } else {
            maxDimension.toFloat() / bitmap.height
        }
        val newWidth = (bitmap.width * scale).toInt().coerceAtLeast(1)
        val newHeight = (bitmap.height * scale).toInt().coerceAtLeast(1)
        val thumbnail = Bitmap.createScaledBitmap(bitmap, newWidth, newHeight, true)
        val output = ByteArrayOutputStream()
        thumbnail.compress(Bitmap.CompressFormat.JPEG, THUMBNAIL_QUALITY, output)
        if (thumbnail != bitmap) thumbnail.recycle()
        bitmap.recycle()
        return output.toByteArray()
    }

    /**
     * Get pixel dimensions of an image.
     */
    fun imageDimensions(imageData: ByteArray): Pair<Int, Int>? {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(imageData, 0, imageData.size, options)
        if (options.outWidth <= 0 || options.outHeight <= 0) return null
        return Pair(options.outWidth, options.outHeight)
    }

    // -- Video Processing --

    /**
     * Extract video metadata: width, height, duration (seconds).
     */
    fun videoMetadata(context: Context, uri: Uri): Triple<Int, Int, Int>? {
        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(context, uri)
            val width = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_WIDTH)?.toIntOrNull() ?: 0
            val height = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_HEIGHT)?.toIntOrNull() ?: 0
            val durationMs = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)?.toLongOrNull() ?: 0
            val rotation = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)?.toIntOrNull() ?: 0
            // Swap width/height for 90/270 rotation
            val (w, h) = if (rotation == 90 || rotation == 270) Pair(height, width) else Pair(width, height)
            Triple(w, h, (durationMs / 1000).toInt())
        } catch (_: Exception) {
            null
        } finally {
            retriever.release()
        }
    }

    /**
     * Generate a thumbnail from the first frame of a video.
     */
    fun generateVideoThumbnail(
        context: Context,
        uri: Uri,
        maxDimension: Int = THUMBNAIL_MAX_DIMENSION
    ): ByteArray? {
        val retriever = MediaMetadataRetriever()
        return try {
            retriever.setDataSource(context, uri)
            val frame = retriever.getFrameAtTime(0, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
                ?: return null
            val scale: Float = if (frame.width > frame.height) {
                maxDimension.toFloat() / frame.width
            } else {
                maxDimension.toFloat() / frame.height
            }
            val newWidth = (frame.width * scale).toInt().coerceAtLeast(1)
            val newHeight = (frame.height * scale).toInt().coerceAtLeast(1)
            val thumbnail = Bitmap.createScaledBitmap(frame, newWidth, newHeight, true)
            val output = ByteArrayOutputStream()
            thumbnail.compress(Bitmap.CompressFormat.JPEG, THUMBNAIL_QUALITY, output)
            if (thumbnail != frame) thumbnail.recycle()
            frame.recycle()
            output.toByteArray()
        } catch (_: Exception) {
            null
        } finally {
            retriever.release()
        }
    }
}

File: clients/ios/StellarChat/StellarChat/Models/MediaCrypto.swift
import AVFoundation
import CryptoKit
import Foundation
import UIKit

/// Per-file AES-256-GCM encryption for media attachments.
/// Each file gets a fresh random key to avoid reusing the group key
/// for large binary payloads, minimizing AES-GCM nonce-collision risk.
enum MediaCrypto {
    /// Maximum compressed image size in bytes (2 MB).
    static let maxImageBytes = 2_000_000
    /// Maximum video size in bytes before encryption (50 MB).
    static let maxVideoBytes = 50_000_000
    /// Maximum thumbnail dimension in points.
    static let thumbnailMaxDimension: CGFloat = 200
    /// Thumbnail JPEG compression quality.
    static let thumbnailQuality: CGFloat = 0.4

    // MARK: - Encrypt / Decrypt

    /// Encrypt data with a fresh random AES-256-GCM key.
    /// Returns `(encryptedBlob, key)` where encryptedBlob = nonce(12) || ciphertext || tag(16).
    static func encryptMedia(_ data: Data) throws -> (encryptedBlob: Data, key: Data) {
        var keyBytes = [UInt8](repeating: 0, count: 32)
        guard SecRandomCopyBytes(kSecRandomDefault, 32, &keyBytes) == errSecSuccess else {
            throw MediaCryptoError.keyGenerationFailed
        }
        let key = SymmetricKey(data: keyBytes)
        let encrypted = try encryptMedia(data, key: key)
        return (encrypted, Data(keyBytes))
    }

    /// Encrypt data with a provided AES-256-GCM key (used for thumbnails sharing the file key).
    /// Returns combined format: nonce(12) || ciphertext || tag(16).
    static func encryptMedia(_ data: Data, key: SymmetricKey) throws -> Data {
        let sealed = try AES.GCM.seal(data, using: key)
        guard let combined = sealed.combined else {
            throw MediaCryptoError.encryptionFailed
        }
        return combined
    }

    /// Convenience overload accepting raw key bytes.
    static func encryptMedia(_ data: Data, key keyData: Data) throws -> Data {
        try encryptMedia(data, key: SymmetricKey(data: keyData))
    }

    /// Decrypt combined-format blob: nonce(12) || ciphertext || tag(16).
    static func decryptMedia(_ encryptedBlob: Data, key keyData: Data) throws -> Data {
        let symmetricKey = SymmetricKey(data: keyData)
        let box = try AES.GCM.SealedBox(combined: encryptedBlob)
        return try AES.GCM.open(box, using: symmetricKey)
    }

    // MARK: - Image Processing

    /// Compress an image to JPEG within `maxBytes`.
    /// Progressively reduces quality until the size limit is met.
    static func compressImage(_ imageData: Data, maxBytes: Int = maxImageBytes) -> Data? {
        guard let image = UIImage(data: imageData) else { return nil }
        var quality: CGFloat = 0.7
        while quality >= 0.1 {
            if let jpeg = image.jpegData(compressionQuality: quality),
               jpeg.count <= maxBytes {
                return jpeg
            }
            quality -= 0.1
        }
        // Last resort: lowest quality
        return image.jpegData(compressionQuality: 0.1)
    }

    /// Generate a small JPEG thumbnail.
    static func generateThumbnail(_ imageData: Data, maxDimension: CGFloat = thumbnailMaxDimension) -> Data? {
        guard let image = UIImage(data: imageData) else { return nil }
        let size = image.size
        let scale: CGFloat
        if size.width > size.height {
            scale = maxDimension / size.width
        } else {
            scale = maxDimension / size.height
        }
        let newSize = CGSize(width: size.width * scale, height: size.height * scale)
        let renderer = UIGraphicsImageRenderer(size: newSize)
        let thumbnail = renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: newSize))
        }
        return thumbnail.jpegData(compressionQuality: thumbnailQuality)
    }

    /// Get pixel dimensions of an image.
    static func imageDimensions(_ imageData: Data) -> (width: Int, height: Int)? {
        guard let image = UIImage(data: imageData) else { return nil }
        return (Int(image.size.width * image.scale), Int(image.size.height * image.scale))
    }

    // MARK: - Video Processing

    /// Extract video metadata: dimensions and duration.
    static func videoMetadata(_ url: URL) async -> (width: Int, height: Int, duration: Int)? {
        let asset = AVURLAsset(url: url)
        guard let track = try? await asset.loadTracks(withMediaType: .video).first else { return nil }
        let size = try? await track.load(.naturalSize)
        let transform = try? await track.load(.preferredTransform)
        let duration = try? await asset.load(.duration)
        guard let size, let duration else { return nil }
        // Apply transform to handle rotated videos
        let transformed = size.applying(transform ?? .identity)
        let w = Int(abs(transformed.width))
        let h = Int(abs(transformed.height))
        let secs = Int(CMTimeGetSeconds(duration))
        return (w, h, secs)
    }

    /// Generate a thumbnail from the first frame of a video.
    static func generateVideoThumbnail(_ url: URL, maxDimension: CGFloat = thumbnailMaxDimension) async -> Data? {
        let asset = AVURLAsset(url: url)
        let generator = AVAssetImageGenerator(asset: asset)
        generator.appliesPreferredTrackTransform = true
        generator.maximumSize = CGSize(width: maxDimension * 2, height: maxDimension * 2)
        guard let cgImage = try? await generator.image(at: .zero).image else { return nil }
        let uiImage = UIImage(cgImage: cgImage)
        // Scale to thumbnail size
        let size = uiImage.size
        let scale: CGFloat = size.width > size.height
            ? maxDimension / size.width
            : maxDimension / size.height
        let newSize = CGSize(width: size.width * scale, height: size.height * scale)
        let renderer = UIGraphicsImageRenderer(size: newSize)
        let thumbnail = renderer.image { _ in
            uiImage.draw(in: CGRect(origin: .zero, size: newSize))
        }
        return thumbnail.jpegData(compressionQuality: thumbnailQuality)
    }

    /// Compress a video to fit within maxBytes using AVAssetExportSession.
    /// Returns the compressed video data, or nil if compression fails.
    static func compressVideo(_ url: URL, maxBytes: Int = maxVideoBytes) async throws -> Data? {
        // Check if original is already small enough
        let attrs = try FileManager.default.attributesOfItem(atPath: url.path)
        let fileSize = attrs[.size] as? Int ?? 0
        if fileSize <= maxBytes {
            return try Data(contentsOf: url)
        }

        // Transcode to medium quality
        let asset = AVURLAsset(url: url)
        guard let session = AVAssetExportSession(asset: asset, presetName: AVAssetExportPresetMediumQuality) else {
            return nil
        }
        let tempURL = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString)
            .appendingPathExtension("mp4")
        session.outputURL = tempURL
        session.outputFileType = .mp4
        await session.export()

        defer { try? FileManager.default.removeItem(at: tempURL) }

        guard session.status == .completed else { return nil }
        let data = try Data(contentsOf: tempURL)
        if data.count > maxBytes {
            // Still too large — try low quality
            guard let lowSession = AVAssetExportSession(asset: asset, presetName: AVAssetExportPresetLowQuality) else {
                return nil
            }
            let lowURL = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString)
                .appendingPathExtension("mp4")
            lowSession.outputURL = lowURL
            lowSession.outputFileType = .mp4
            await lowSession.export()
            defer { try? FileManager.default.removeItem(at: lowURL) }
            guard lowSession.status == .completed else { return nil }
            let lowData = try Data(contentsOf: lowURL)
            return lowData.count <= maxBytes ? lowData : nil
        }
        return data
    }
}

enum MediaCryptoError: LocalizedError {
    case keyGenerationFailed
    case encryptionFailed

    var errorDescription: String? {
        switch self {
        case .keyGenerationFailed: return "Failed to generate encryption key"
        case .encryptionFailed: return "Failed to encrypt media"
        }
    }
}
